eftest/src/eftest/runner.clj

```diff
@@ -119,15 +126,25 @@
       (test/do-report {:type :end-test-ns, :ns ns})
       @test/*report-counters*)))
 
+(defn- arrange-namespaces [seed grouped-namespaces]
+  (if seed
+    (reduce
+     (fn [arranged namespace]
+       (conj arranged [namespace (get grouped-namespaces namespace)]))
+     []
+     (deterministic-shuffle (sort-by #(.getName %) (keys grouped-namespaces))  seed))
+    grouped-namespaces))
+
 (defn- test-all [vars {:as opts
-                       :keys [capture-output?]
-                       :or {capture-output? true}}]
+                       :keys [capture-output? randomize-seed]
+                       :or {capture-output? true randomize-seed nil}}]
   (let [report   (synchronize test/report)
         executor (delay (Executors/newCachedThreadPool))
         mapf     (if (multithread-namespaces? opts)
                    (partial pmap* @executor)
                    map)
         f        #(->> (group-by (comp :ns meta) vars)
+                       (arrange-namespaces randomize-seed)
```

Comment:
This can be written without `arrange-namespaces`, I think, as we can just shuffle the key/value pairs directly. So instead:

```clojure
(cond->> randomize-seed (deterministic-shuffle randomize-seed))
```

And swap around the arguments of `deterministic-shuffle`:

```clojure
(defn- deterministic-shuffle [seed ^java.util.Collection coll]
  (let [al (java.util.ArrayList. coll)
        rng (java.util.Random. seed)]
    (java.util.Collections/shuffle al rng)
    (vec al)))
```

Reply:
Shuffling is index based, which means that it produces different results depending on the order of the elements in input coll. `group-by` returns a hashmap, and hashmaps are somewhat randomly ordered. This means that shuffling will produce different results for the same set of namespaces, i.e. it won't be deterministic.
I managed to simplify the function though.

Reply:
That's a good point. Then what about adding a `(sort-by key)`, e.g.

```clojure
(->> (group-by (comp :ns meta) vars)
     (sort-by key)
     (cond->> randomize-seed (deterministic-shuffle randomize-seed))
     ...)

Reply:
That'll change the default behavior, namespaces will always be sorted.
Also, it should be `(sort-by #(-> % first .getName))` because keys are `clojure.lang.Namespace`'s after grouping.

Reply:
Ah, good point. I'd prefer `str` instead of `.getName` and `key` instead of `first`. So:

```clojure
(sort-by (comp str key))
```

I also don't mind the namespaces being sorted by default. Having deterministic test ordering by default doesn't seem like a bad idea.

Reply:
I removed `arrange-namespaces` and added `sort-by`. We cannot use `cond->>` as shown above though because it's inside `->>` and sorted vars are passed as the last argument to `cond->>`.

We can simplify it further. Since we're always sorting by default, we can always do deterministic shuffling with `randomize-seed` set to 0 by default. What do you think?

Also, maybe a next step could be a feature that automatically generates a seed, prints it out (so tests run can be reproduced later) and sets `randomize-seed`. Some boolean flag, like `random-order?`.

Reply:
Yes, that `fn` inside the `->>` sure looks ugly, but I can't think of a better way without changing the behaviour or adding a new function.

I like the idea of setting `randomize-seed` to `0` by default. That makes the code a lot simpler.

Reply:
Ok, done.